File: backend/app/routers/finance.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any
from uuid import UUID
from decimal import Decimal

from ..core.database import get_db_pool
from ..middleware.auth_middleware import get_current_user, AuthenticatedUser
from ..services.ledger_service import LedgerService, VerificationStatus
from ..services.deal_service import DealService

router = APIRouter(prefix="/deals/{deal_id}/ledger", tags=["Finance"])


class VerificationRequest(BaseModel):
    entry_id: UUID
    status: VerificationStatus
    notes: Optional[str] = None


class AdminAdjustmentRequest(BaseModel):
    amount: Decimal
    direction: str = Field(..., pattern="^(CREDIT|DEBIT|INFO)$")
    description: str
    metadata: Optional[Dict[str, Any]] = {}


def _require_admin(current_user: AuthenticatedUser) -> None:
    if not any(role in (current_user.roles or []) for role in ["ADMIN", "CEO"]):
        raise HTTPException(status_code=403, detail="Admin access required")


async def _ensure_deal_access(pool, deal_id: UUID, current_user: AuthenticatedUser) -> None:
    deal_service = DealService(pool)
    access_check = await deal_service.get_deal(deal_id=deal_id, user_id=current_user.user_id)

    if not access_check.get("success"):
        error_msg = (access_check.get("error") or "").lower()
        if "not found" in error_msg:
            raise HTTPException(status_code=404, detail=access_check.get("error", "Deal not found"))
        if "denied" in error_msg:
            raise HTTPException(status_code=403, detail=access_check.get("error", "Access denied"))
        raise HTTPException(status_code=400, detail=access_check.get("error", "Invalid deal access"))


@router.get("")
async def get_deal_ledger(
    deal_id: UUID,
    current_user: AuthenticatedUser = Depends(get_current_user),
):
    """
    Get financial ledger for a deal.
    Accessible only by deal participants and admins.
    """
    pool = get_db_pool()
    await _ensure_deal_access(pool, deal_id, current_user)

    service = LedgerService(pool)
    result = await service.get_ledger(deal_id)

    if not result.get("success"):
        raise HTTPException(status_code=404, detail=result.get("error", "Ledger not found"))

    return result


@router.post("/verify-booking")
async def verify_booking_declaration(
    deal_id: UUID,
    data: VerificationRequest,
    current_user: AuthenticatedUser = Depends(get_current_user),
):
    """
    Agent/Admin verifies payment-related ledger entries.
    """
    if not any(role in (current_user.roles or []) for role in ["AGENT", "ADMIN", "CEO"]):
        raise HTTPException(status_code=403, detail="Agent/Admin access required")

    pool = get_db_pool()
    await _ensure_deal_access(pool, deal_id, current_user)

    service = LedgerService(pool)
    result = await service.verify_entry(
        entry_id=data.entry_id,
        verifier_id=current_user.user_id,
        status=data.status,
        notes=data.notes,
        deal_id=deal_id,
    )

    if not result.get("success"):
        error_msg = (result.get("error") or "").lower()
        if "not found" in error_msg:
            raise HTTPException(status_code=404, detail=result.get("error", "Entry not found"))
        if "not authorized" in error_msg:
            raise HTTPException(status_code=403, detail=result.get("error", "Not authorized"))
        raise HTTPException(status_code=400, detail=result.get("error", "Failed to verify entry"))

    return result


@router.post("/override")
async def admin_override_ledger(
    deal_id: UUID,
    data: AdminAdjustmentRequest,
    current_user: AuthenticatedUser = Depends(get_current_user),
):
    """
    Admin-only: Force add an adjustment entry.
    """
    _require_admin(current_user)

    pool = get_db_pool()
    await _ensure_deal_access(pool, deal_id, current_user)

    service = LedgerService(pool)
    result = await service.record_entry(
        deal_id=deal_id,
        entry_type="ADJUSTMENT",
        amount=data.amount,
        direction=data.direction,
        description=data.description,
        user_id=current_user.user_id,
        metadata=data.metadata,
        verification_status="VERIFIED",  # Admin entries are auto-verified
    )

    if not result.get("success"):
        raise HTTPException(status_code=400, detail=result.get("error", "Failed to record entry"))

    return result

File: frontend/src/lib/api/finance.ts
import * as api from '@/lib/api';
import { LedgerResponse, VerificationStatus, CommissionStatusResponse } from '../types/finance';

export async function getDealLedger(dealId: string): Promise<LedgerResponse> {
    try {
        const response = await api.get(`/deals/${dealId}/ledger`);
        // api.get returns parsed JSON directly (no .data wrapper)
        return response as LedgerResponse;
    } catch (error: any) {
        // Ledger may not exist yet for early-stage deals — graceful degradation
        if (error?.status === 404) {
            return { success: false, ledger: null } as any;
        }
        throw error;
    }
}

export async function verifyBookingDeclaration(
    dealId: string,
    entryId: string,
    status: VerificationStatus,
    notes?: string
): Promise<{ success: boolean; status: string }> {
    const response = await api.post(`/deals/${dealId}/ledger/verify-booking`, {
        entry_id: entryId,
        status,
        notes,
    });
    return response.data;
}

export async function adminOverrideLedger(
    dealId: string,
    amount: number,
    direction: 'CREDIT' | 'DEBIT' | 'INFO',
    description: string,
    metadata?: Record<string, unknown>
): Promise<{ success: boolean; entry_id: string }> {
    const response = await api.post(`/deals/${dealId}/ledger/override`, {
        amount,
        direction,
        description,
        metadata,
    });
    return response.data;
}

// ============================================================================
// PHASE 5A: Commission Lifecycle & Enhanced Payment Recording
// ============================================================================

export async function getCommissionStatus(dealId: string): Promise<CommissionStatusResponse> {
    const response = await api.get(`/deals/${dealId}/commission`);
    return response.data;
}

export async function authorizeCommission(
    dealId: string,
    notes?: string
): Promise<{ success: boolean; message: string; commission_status: string }> {
    const response = await api.post(`/deals/${dealId}/commission/authorize`, {
        notes,
    });
    return response.data;
}

export async function recordPaymentWithProof(
    dealId: string,
    data: {
        entry_type: string;
        amount: number;
        description: string;
        proof_url: string;
        payment_method: string;
        bank_reference?: string;
        notes?: string;
    }
): Promise<{ success: boolean; entry_id: string; message: string }> {
    const response = await api.post(`/deals/${dealId}/ledger/payment`, data);
    return response.data;
}

export async function confirmPaymentEntry(
    dealId: string,
    entryId: string
): Promise<{ success: boolean; message: string }> {
    const response = await api.post(`/deals/${dealId}/ledger/confirm`, {
        entry_id: entryId,
    });
    return response.data;
}
